File: employee_imp.cpp
TC:O(N)
SC:O(N)


/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
private:
    int total_imp = 0;
public:
    int getImportance(vector<Employee*> employees, int id) {
        unordered_map<int,Employee*>umap;
        for(auto emp:employees){
            umap[emp->id] = emp;
        }
        queue<int>q;
        // to start bfs we need one starting point
        q.push(id);
        while(!q.empty()){
            int emp_id = q.front();q.pop();
            vector<int>sub = umap[emp_id]->subordinates;
            total_imp +=umap[emp_id]->importance;
            for(int s:sub){
                q.push(s);
            }   
        }
        return total_imp;
        
        
    }
};

File: EmployeeImp.py
#Time complexity: O(n)
#Space complexity: O(h)
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    dict1={}
    imp=0
    def getImportance(self, employees, id: int) -> int:
        self.dict1={}
        self.imp=0
        for e in employees:
            self.dict1[e.id]=e
        self.dfs(id)
        return self.imp
        
    def dfs(self,root):
        if not root:
            return
        self.imp+= self.dict1[root].importance
        for sub in self.dict1[root].subordinates:
            self.dfs(sub)
        

        
        
        